### PR TITLE
Validate JWT 'typ' header.

### PR DESCRIPTION
**Pull request by maxlambrecht**
Validates that in case the `typ` header is present, it is either `JWT` or `JOSE`.

Addresses #58 

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

### CODE REVIEW

File: java-spiffe-core/src/test/java/io/spiffe/svid/jwtsvid/JwtSvidParseAndValidateTest.java [hunk 1]
```diff
@@ -119,119 +131,159 @@ void testParseAndValidate_nullAudience_throwsNullPointerException() throws JwtSv
...
+        KeyPair key3 = TestUtils.generateRSAKeyPair(2048);
+
+        TrustDomain trustDomain = TrustDomain.of("test.domain");
+        JwtBundle jwtBundle = new JwtBundle(trustDomain);
+        jwtBundle.putJwtAuthority("authority1", key1.getPublic());
+        jwtBundle.putJwtAuthority("authority2", key2.getPublic());
+        jwtBundle.putJwtAuthority("authority3", key3.getPublic());
+
+        SpiffeId spiffeId = trustDomain.newSpiffeId("host");
+        Date expiration = new Date(System.currentTimeMillis() + 3600000);
```
Comment (rturner3) [inline java-spiffe-core/src/test/java/io/spiffe/svid/jwtsvid/JwtSvidParseAndValidateTest.java:184]: nit: express the constant time in the future as a multiplication so it's a little easier to understand the value
```suggestion
        Date expiration = new Date(System.currentTimeMillis() + (60 * 60 * 1000));
```

File: java-spiffe-core/src/test/java/io/spiffe/svid/jwtsvid/JwtSvidParseAndValidateTest.java [hunk 2]
```diff
@@ -119,119 +131,159 @@ void testParseAndValidate_nullAudience_throwsNullPointerException() throws JwtSv
...
+        Set<String> audience = new HashSet<String>() {{add("audience1"); add("audience2");}};
```
Comment (rturner3) [inline java-spiffe-core/src/test/java/io/spiffe/svid/jwtsvid/JwtSvidParseAndValidateTest.java:185]: nit: `s/HashSet<String>/HashSet<>/`

File: java-spiffe-core/src/test/java/io/spiffe/svid/jwtsvid/JwtSvidParseAndValidateTest.java [hunk 3]
```diff
@@ -119,119 +131,159 @@ void testParseAndValidate_nullAudience_throwsNullPointerException() throws JwtSv
 
         return Stream.of(
                 Arguments.of(TestCase.builder()
-                        .name("1. success using EC signature")
+                        .name("1. using EC signature")
```
Comment (rturner3) [inline java-spiffe-core/src/test/java/io/spiffe/svid/jwtsvid/JwtSvidParseAndValidateTest.java:134]: Consider just removing the numbers from all these test names since they inevitably change whenever test cases are added/removed. It results in a lot of noise in diffs and manual work to fix the test name numbers. I don't really see too much value in building in ordering into the names.

This can also be done as a separate PR since I see this convention is already adopted in many of the tests.

File: java-spiffe-core/src/main/java/io/spiffe/svid/jwtsvid/JwtSvid.java [hunk 4]
```diff
@@ -290,7 +300,7 @@ private static void validateAudience(final List<String> audClaim, final Set<Stri
 
     private static JwtSignatureAlgorithm parseAlgorithm(JWSAlgorithm algorithm) throws JwtSvidException {
         if (algorithm == null) {
-            throw new JwtSvidException("jwt header 'alg' is required");
+            throw new JwtSvidException("JWT header 'alg' is required");
```
[the remark this answers is not in the document]
Comment (maxlambrecht) [reply java-spiffe-core/src/main/java/io/spiffe/svid/jwtsvid/JwtSvid.java:303]: it's unrelated, I make that amendment to keep consistency in that file using capitalized 'JWT', and as it doesn't break any functionality I think it's not worth a separate PR to do it. But if you still think it would be better to move to separate PR I can do it.

File: java-spiffe-core/src/test/java/io/spiffe/svid/jwtsvid/JwtSvidParseAndValidateTest.java (hunk 3 above)
Comment (maxlambrecht) [reply java-spiffe-core/src/test/java/io/spiffe/svid/jwtsvid/JwtSvidParseAndValidateTest.java:134]: I think you are right, it helped when debugging failing tests, but it produces noise and manual work when adding new tests. I'll remove them in these two test classes I touched in this PR and will address the rest in a separate PR.

**Review by rturner3 — APPROVED** [review]
